src/Libraries/CoreNodes/String.cs

```diff
@@ -429,9 +430,34 @@ public static string Remove(string str, int startIndex, int? count = null)
 
             if (_count < 0)
             {
-                startIndex += _count;
+                // The input count is negative means the removal process is required operate from right to left.
+                // However, the removal process in this function always operates rightwards (left to right).
+                // Therefore, a conversion for start index needs to be done in order to change
+                // from leftwards removal (right to left) to rightwards removal (left to right).
+                startIndex = startIndex + _count + 1; 
                 _count *= -1;
             }
+            
+            if(_count > str.Length)
+            {
+                throw new ArgumentOutOfRangeException("count", "Count is out of range!");
+            }
+
+            if (startIndex >= str.Length || startIndex < 0) 
+            {
+                // startIndex of an array must be within the string length. 
+                // If after the conversion of negative startIndex, startIndex is still
+                // less than zero, ArgumentOutOfRangeException is occured.
+                if (startIndex == str.Length && str.Length == 0)
+                {
+                    return str.Remove(startIndex, _count);
```

Comment:
@ke-yu In the original test case, there is an edge case test: Assert.AreEqual("", String.Remove("", 0));
So I was thinking of just preserving that so we still allow str.Length is 0.